index.js

```diff
@@ -20,3 +32,110 @@ export async function * linkdex (carStream) {
     }
   }
 }
+
+/**
+ * A util to record all the links from a given set of Blocks.
+ * Pass blocks to `decodeAndIndex` as you iterate over a CAR.
+ * Then call `getDagStructureLabel` to find out if the dag is
+ * complete... if all linked to CIDs are also contained in in
+ * the set, then it is complete.
+ */
+export class LinkIndexer {
+  constructor () {
+    /**
+     * Map block CID to the set of CIDs it links to
+     * @type {Map<string, Set<string>>}
+     * */
+    this.idx = new Map()
+    this.indexed = 0
+    this.undecodable = 0
+  }
+
+  /**
+   * Decode the block and index any CIDs it links to
+   * @param {import('@ipld/car/api').Block} block
+   * @param {object} [opts]
+   * @param {object} [opts.codecs]
+   */
+  decodeAndIndex ({ cid, bytes }, opts) {
+    const block = maybeDecode({ cid, bytes }, opts)
+    if (!block) {
+      this.undecodable++
+    } else {
+      this._index(block)
+    }
+  }
+
+  /**
+   * Index all the links from the block
+   * @param {import('multiformats/block').Block} block
+   */
+  _index (block) {
+    this.indexed++
```

Comment:
Should this be after next condition where we validation if given block was already indexed?

Reply:
I'm using indexed to count how may blocks this thing was asked to index... so a total block count. Later we use the map size to find the unique block count.